File: app/src/main/java/com/gameautopilot/app/brain/OpenAiCompatibleBrain.kt
package com.gameautopilot.app.brain

import java.io.IOException
import java.util.concurrent.TimeUnit
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject

class OpenAiCompatibleBrain(
    private val baseUrl: String,
    private val apiKey: String,
    private val model: String,
    private val client: OkHttpClient = defaultClient
) : Brain {

    override suspend fun decide(ctx: BrainContext): BrainDecision {
        if (apiKey.isBlank()) throw BrainException("API key not set")

        val body = buildRequestBody(ctx)
        val req = Request.Builder()
            .url("${baseUrl.trimEnd('/')}/chat/completions")
            .addHeader("Authorization", "Bearer $apiKey")
            .addHeader("Content-Type", "application/json")
            .post(body.toString().toRequestBody(JSON))
            .build()

        val raw = try {
            client.newCall(req).await()
        } catch (e: IOException) {
            throw BrainException("Network error: ${e.message}", e)
        }

        raw.use { resp ->
            val text = resp.body?.string().orEmpty()
            if (!resp.isSuccessful) {
                throw BrainException("HTTP ${resp.code}: ${text.take(400)}")
            }
            return parseDecision(text)
        }
    }

    private fun buildRequestBody(ctx: BrainContext): JSONObject {
        val systemMsg = JSONObject().apply {
            put("role", "system")
            put("content", PromptBuilder.systemPrompt(ctx))
        }
        val userMsg = JSONObject().apply {
            put("role", "user")
            put("content", JSONArray().apply {
                put(JSONObject().apply {
                    put("type", "text")
                    put("text", PromptBuilder.userText(ctx))
                })
                put(JSONObject().apply {
                    put("type", "image_url")
                    put("image_url", JSONObject().apply {
                        put("url", "data:image/jpeg;base64,${ctx.screenshotBase64Jpeg}")
                    })
                })
            })
        }

        return JSONObject().apply {
            put("model", model)
            put("temperature", 0.2)
            put("max_tokens", 800)
            put("response_format", JSONObject().apply { put("type", "json_object") })
            put("messages", JSONArray().apply {
                put(systemMsg)
                put(userMsg)
            })
        }
    }

    private fun parseDecision(httpBody: String): BrainDecision {
        val outer = JSONObject(httpBody)
        val choices = outer.optJSONArray("choices") ?: throw BrainException("No choices in response")
        if (choices.length() == 0) throw BrainException("Empty choices")
        val content = choices.getJSONObject(0)
            .optJSONObject("message")
            ?.optString("content")
            ?: throw BrainException("No content in first choice")
        return BrainResponseParser.parse(content)
    }

    companion object {
        private val JSON = "application/json; charset=utf-8".toMediaType()

        val defaultClient: OkHttpClient by lazy {
            OkHttpClient.Builder()
                .connectTimeout(15, TimeUnit.SECONDS)
                .readTimeout(60, TimeUnit.SECONDS)
                .writeTimeout(60, TimeUnit.SECONDS)
                .build()
        }
    }
}

File: mac/Sources/GameAutopilotMac/Brain/OpenAICompatibleBrain.swift
import Foundation

/// OpenAI-compatible `/chat/completions` brain -- used for both the OpenAI
/// and NVIDIA NIM providers (NIM speaks the same request/response shape).
/// Mirrors OpenAiCompatibleBrain.kt with URLSession instead of OkHttp and
/// JSONSerialization instead of org.json.
final class OpenAICompatibleBrain: Brain {
    private let baseUrl: String
    private let apiKey: String
    private let model: String
    private let session: URLSession

    init(baseUrl: String, apiKey: String, model: String, session: URLSession = .shared) {
        self.baseUrl = baseUrl
        self.apiKey = apiKey
        self.model = model
        self.session = session
    }

    func decide(_ ctx: BrainContext) async throws -> BrainDecision {
        guard !apiKey.isEmpty else { throw BrainError("API key not set") }

        let trimmedBase = baseUrl.hasSuffix("/") ? String(baseUrl.dropLast()) : baseUrl
        guard let url = URL(string: "\(trimmedBase)/chat/completions") else {
            throw BrainError("Invalid base URL")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: buildRequestBody(ctx))

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw BrainError("Network error: \(error.localizedDescription)")
        }

        guard let http = response as? HTTPURLResponse else {
            throw BrainError("No HTTP response")
        }
        guard (200..<300).contains(http.statusCode) else {
            let text = String(data: data, encoding: .utf8) ?? ""
            throw BrainError("HTTP \(http.statusCode): \(text.prefix(400))")
        }

        return try parseDecision(data)
    }

    private func buildRequestBody(_ ctx: BrainContext) -> [String: Any] {
        let systemMsg: [String: Any] = ["role": "system", "content": PromptBuilder.systemPrompt(ctx)]
        let userMsg: [String: Any] = [
            "role": "user",
            "content": [
                ["type": "text", "text": PromptBuilder.userText(ctx)],
                ["type": "image_url", "image_url": ["url": "data:image/jpeg;base64,\(ctx.screenshotBase64Jpeg)"]]
            ]
        ]
        return [
            "model": model,
            "temperature": 0.2,
            "max_tokens": 800,
            "response_format": ["type": "json_object"],
            "messages": [systemMsg, userMsg]
        ]
    }

    private func parseDecision(_ data: Data) throws -> BrainDecision {
        guard let outer = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw BrainError("Malformed response body")
        }
        guard let choices = outer["choices"] as? [[String: Any]], !choices.isEmpty else {
            throw BrainError("No choices in response")
        }
        guard let message = choices[0]["message"] as? [String: Any],
              let content = message["content"] as? String
        else {
            throw BrainError("No content in first choice")
        }
        return try BrainResponseParser.parse(content)
    }
}
